compiler/damlc/lib/DA/Cli/Damlc/Test.hs

```diff
@@ -73,24 +76,36 @@ testRun h inFiles lfVersion coverage color mbJUnitOutput  = do
     let files = nubOrd $ concat $ inFiles : catMaybes deps
 
     results <- runActionSync h $ do
-        dalfs <- Shake.forP files dalfForScenario
         Shake.forP files $ \file -> do
+            dalf <- dalfForScenario file
```

Comment:
I’m a bit confused. I was expecting us to change which tests we run based on `runAllTests` but that doesn’t seem to be the case. We always run tests against all modules in the main dalf. What we change is whether coverage is reported relative to the main dalf or relative to all dalfs.

I see good usecases for the following configurations:
1. Run all scenarios in the main dalf, report coverage against main dalf. You’ll mainly use that for performance reasons. Running scenarios against all dalfs can clearly be slower.
2. Run all scenarios in all dalfs, report coverage against all dalfs. Doesn’t need much of an explanation I think. You run as much as possible and report it against everything it depends on.

The current `--all` flag implements a third option:

3. Run scenarios of the main dalf and report coverage against all dalfs.

This seems a bit of a weird configuration. You can’t expect to get 100% in that setting without repeating potential tests in your dependencies. If your dependencies don’t have any tests, this is equivalent to 2 so not very useful as a separate configuration. The one usecase I can see is if your dependencies do contain tests but for performance reasons you don’t want to run them and you’re only looking at relative improvements. I think this is a valid usecase but I think it points at a more general feature that might solve this and others:

A way to only run specific tests. Filtering to the main dalf is one specific instance of that but I think there are more, here’s a very rough idea of a syntax for filtering:

1. pkgid:modulename:testname runs exactly that test
2. packagename:modulename:testname runs exactly that test, errors out if package name is ambigous
3. packagename-version:modulename:testname runs exactly that test, errors out if package name & version is ambigous
4. :modulename:testname runs exactly that test in the main dalf
5. :modulename: runs all tests in that module.

You probably also want some form of pattern matching (e.g., regex) in all 3 positions.

That’s definitely not something for this PR and I don’t think anybody has asked for it so far so I would recommend to adjust this PR to switch between 1 and 2 and open a separate issue for 3.

Reply:
@cocreature I think the source of my confusion comes from line 75-76 in Damlc/Test.hs. I was working under the assumption that this will give me all dependencies, not just the main dalf, and hence would implement option 1/2 depending on the flag set or not. Is that not the case?

Reply:
`dalfForScenario` returns a dalf for the given file based on files from the current project so a subset of what `daml build` gives you (the reason why it’s a subset is that we want to run tests even if other files fail to compile at least in daml studio). `files` refers to the files in the current project. `getExternalPackages` provides you with the dependencies.

Reply:
@cocreature I tested this manually. A test would probably have to go to the integration tests for creating and importing an external package, not sure if we should do it since those are fairly slow already.

Reply:
@robin-da you don’t need the integration tests for building code that depends on an external package. We already have a code that does this, see https://github.com/digital-asset/daml/blob/4fcd2b833f40c1c4a4d92b45b87380d9514e85ba/compiler/damlc/tests/src/DamlcTest.hs#L273.